src/service/report-service.ts

```diff
@@ -0,0 +1,388 @@
+      .innerJoin(Transaction, 'transaction', 'transaction.id = subTransaction.transactionId');
+    this.addSelectTotals(query);
+
+    const data = await this.addSubTransactionRowFilter(query, forId, fromDate, tillDate)
+      .getRawAndEntities();
+
+    if (data.raw.length > 1) throw new Error('Multiple results when getting totals');
+
+    const totalExclVat = asDinero(data.raw[0].total_excl_vat || 0);
+    const totalInclVat = asDinero(data.raw[0].total_incl_vat || 0);
+    return {
+      totalExclVat,
+      totalInclVat,
+    };
+  }
+
+  /**
+   * Gets the report for the given user
+   * @param parameters - The parameters to get the report for
+   * @returns {Promise<Report>} - The report
+   */
+  public async getReport(parameters: ReportParameters): Promise<Report> {
+    const { fromDate, tillDate, forId } = parameters;
+    const productEntries = await this.getProductEntries(forId, fromDate, tillDate);
+    const vatEntries = await this.getVatEntries(forId, fromDate, tillDate);
+    const categoryEntries = await this.getCategoryEntries(forId, fromDate, tillDate);
+    const getPosEntries = await this.getPosEntries(forId, fromDate, tillDate);
+    const getContainerEntries = await this.getContainerEntries(forId, fromDate, tillDate);
+    const totals = await this.getTotals(forId, fromDate, tillDate);
+
+    return {
+      forId,
+      fromDate,
+      tillDate,
+      data: {
+        products: productEntries,
+        vat: vatEntries,
+        categories: categoryEntries,
+        pos: getPosEntries,
+        containers: getContainerEntries,
+      },
+      totalExclVat: totals.totalExclVat,
+      totalInclVat: totals.totalInclVat,
+    };
+  }
+}
+
+export class SalesReportService extends ReportService {
+
+  protected addSubTransactionRowFilter<T>(query: SelectQueryBuilder<T>, forId: number, fromDate: Date, tillDate: Date): SelectQueryBuilder<T> {
+    return query
+      .where('subTransaction.toId = :userId', { userId: forId })
+      .andWhere('subTransaction.createdAt >= :fromDate', { fromDate: toMySQLString(fromDate) })
+      .andWhere('subTransaction.createdAt < :tillDate', { tillDate: toMySQLString(tillDate) });
+  }
+
+  getReport(parameters: ReportParameters): Promise<SalesReport> {
+    return super.getReport(parameters);
+  }
+}
+
+export class BuyerReportService extends ReportService {
+
+  protected addSubTransactionRowFilter<T>(query: SelectQueryBuilder<T>, forId: number, fromDate: Date, tillDate: Date): SelectQueryBuilder<T> {
+    return query
+      .where('transaction.fromId = :userId', { userId: forId })
+      .andWhere('subTransaction.createdAt >= :fromDate', { fromDate: toMySQLString(fromDate) })
+      .andWhere('subTransaction.createdAt < :tillDate', { tillDate: toMySQLString(tillDate) });
+  }
+
+  getReport(parameters: ReportParameters): Promise<BuyerReport> {
+    return super.getReport(parameters);
+  }
+}
```

Comment:
I really liked the more single responsibility and encapsulation of the seller vs purchase classes. I could change it, but that does make it harder down the line to see if a report is of type seller or of type purchase. This is useful for stuff like pdf reporting and just makes the whole thing more explicit.

Reply:
Let's keep it this way for now, we can discuss stuff like this later. Also because I think we differ in view as to what we should use the ReportService for.